[Docs] Improvements to custom flavor and custom orchestrator pages (#1747)
zenml-io/zenml: docs/book/user-guide/advanced-guide/custom-stack-solutions/implement-a-custom-stack-component.md

> # Implement a custom stack component

Reviewer: See here it's 'implement'. So maybe in the TOC it can be that as well?

Reviewer 2: H1 headers aren't shown so doesn't matter but yeah I also like implement best 👍

When building a sophisticated MLOps Platform, you will often need to come up 
with custom-tailored solutions for your infrastructure or tooling. 

> ZenML is built on the philosophy of composability and reusability which is why

Reviewer: ```suggestion
ZenML is built around the values of composability and reusability which is why
```

flavor below:

```python
from zenml.enums import StackComponentType
from zenml.stack import Flavor

class LocalArtifactStore(BaseArtifactStore):
    ...

class LocalArtifactStoreConfig(BaseArtifactStoreConfig):
    ...

class LocalArtifactStoreFlavor(Flavor):

    @property
    def name(self) -> str:
        """Returns the name of the flavor."""
        return "local"

    @property
    def type(self) -> StackComponentType:
        """Returns the flavor type."""
        return StackComponentType.ARTIFACT_STORE

    @property
    def config_class(self) -> Type[LocalArtifactStoreConfig]:
        """Config class of this flavor."""
        return LocalArtifactStoreConfig

    @property
    def implementation_class(self) -> Type[LocalArtifactStore]:
        """Implementation class of this flavor."""
        return LocalArtifactStore
```

See the full code of the base `Flavor` class definition
[here](https://github.com/zenml-io/zenml/blob/main/src/zenml/stack/flavor.py#L29).

## Implementing a Custom Stack Component Flavor

Let's recap what we just learned by reimplementing the `S3ArtifactStore` from 
the `aws` integration as a custom flavor.

> Let's start with the configuration class. Here we need to define the 

Reviewer: ```suggestion
We can start with the configuration class: here we need to define the 
```

`SUPPORTED_SCHEMES` class variable introduced by the `BaseArtifactStore`. We 
also define several additional configuration values that users can use to 

> configure how the artifact store will authenticate to AWS:

Reviewer: ```suggestion
configure how the artifact store will authenticate with AWS:
```

```

{% hint style="info" %}
For flavors that require additional dependencies, you should make sure to 
define your implementation, config, and flavor classes in separate Python files 
and to only import the implementation class inside the `implementation_class` 
property of the flavor class. Otherwise ZenML will not be able to load and 
validate your flavor configuration without the dependencies installed.
{% endhint %}

## Managing a Custom Stack Component Flavor

Once you have defined your implementation, config, and flavor classes, you can 
register your new flavor through the ZenML CLI:

```shell
zenml artifact-store flavor register <path.to.MyS3ArtifactStoreFlavor>
```

{% hint style="info" %}
Make sure to point to the flavor class via dot notation!
{% endhint %}

For example, if your flavor class `MyS3ArtifactStoreFlavor` is defined in 
`flavors/my_flavor.py`, you'd register it by doing:

```shell
zenml artifact-store flavor register flavors.my_flavor.MyS3ArtifactStoreFlavor
```

Afterwards, you should see the new custom artifact store flavor in the list of 
available artifact store flavors:

```shell
zenml artifact-store flavor list
```

> And that's it, you now have a custom stack component flavor that you can use in 

Reviewer: ```suggestion
And that's it! You now have a custom stack component flavor that you can use in 
```

* You can use the ZenML CLI to find which exact configuration values a specific
flavor requires. Check out 
[this 3-minute video](https://www.youtube.com/watch?v=CQRVSKbBjtQ) for more 
information.

* You can keep changing the `Config` and `Settings` of your flavor after
registration. ZenML will pick up these "live" changes when running pipelines.

* Note that changing the config in a breaking way requires an update of the
component (not a flavor). E.g., adding a mandatory name to flavor X field
will break a registered component of that flavor. This may lead to a
completely broken state where one should delete the component and
re-register it.

* Always test your flavor thoroughly before using it in production. Make sure
it works as expected and handles errors gracefully.

* Keep your flavor code clean and well-documented. This will make it easier
for others to use and contribute to your flavor.

* Follow best practices for the language and libraries you're using. This will
help ensure your flavor is efficient, reliable, and easy to maintain.

* We recommend you develop new flavors by using existing flavors as a reference.
A good starting point are the flavors defined in the

> [officially ZenML intetgrations](https://github.com/zenml-io/zenml/tree/main/src/zenml/integrations).

Reviewer: ```suggestion
[official ZenML integrations](https://github.com/zenml-io/zenml/tree/main/src/zenml/integrations).
```
